### PR TITLE
Builds: index on jobset_id, largely join by jobset_id

### PR DESCRIPTION
**Pull request by grahamc**
This change is primarily focused on improving the jobset overview query. This change reduces its best case from about 5 seconds to less than one second.

Before:

```
hydra=# explain (analyze, timing) SELECT me.name, me.id, me.project, me.description, me.nixexprinput, me.nixexprpath, me.errormsg, me.errortime, me.lastcheckedtime, me.triggertime, me.enabled, me.enableemail, me.hidden, me.emailoverride, me.keepnr, me.checkinterval, me.schedulingshares, me.fetcherrormsg, me.forceeval, me.starttime, me.type, me.flake, (select count(*) from Builds as a where a.finished = 0 and me.project = a.project and me.name = a.jobset and a.isCurrent = 1), (select count(*) from Builds as a where a.finished = 1 and me.project = a.project and me.name = a.jobset and buildstatus <> 0 and a.isCurrent = 1), (select count(*) from Builds as a where a.finished = 1 and me.project = a.project and me.name = a.jobset and buildstatus = 0 and a.isCurrent = 1), (select count(*) from Builds as a where me.project = a.project and me.name = a.jobset and a.isCurrent = 1) FROM jobsets me WHERE ( me.project = 'nixpkgs' ) ORDER BY hidden ASC, enabled DESC, name;
                                                                                 QUERY PLAN                                                                                 
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Sort  (cost=231579.80..231580.04 rows=95 width=534) (actual time=5162.712..5162.720 rows=95 loops=1)
   Sort Key: me.hidden, me.enabled DESC, me.name
   Sort Method: quicksort  Memory: 60kB
   ->  Bitmap Heap Scan on jobsets me  (cost=5.41..231576.68 rows=95 width=534) (actual time=29.194..5162.401 rows=95 loops=1)
         Recheck Cond: (project = 'nixpkgs'::text)
         Heap Blocks: exact=45
         ->  Bitmap Index Scan on jobsets_pkey  (cost=0.00..5.39 rows=95 width=0) (actual time=0.020..0.021 rows=95 loops=1)
               Index Cond: (project = 'nixpkgs'::text)
         SubPlan 1
           ->  Aggregate  (cost=454.36..454.37 rows=1 width=8) (actual time=10.805..10.805 rows=1 loops=95)
                 ->  Bitmap Heap Scan on builds a  (cost=453.23..454.35 rows=1 width=0) (actual time=10.562..10.764 rows=964 loops=95)
                       Recheck Cond: ((me.project = project) AND (me.name = jobset) AND (iscurrent = 1) AND (finished = 0))
                       Heap Blocks: exact=15868
                       ->  BitmapAnd  (cost=453.23..453.23 rows=1 width=0) (actual time=10.531..10.531 rows=0 loops=95)
                             ->  Bitmap Index Scan on indexbuildsonjobsetiscurrent  (cost=0.00..75.61 rows=905 width=0) (actual time=4.442..4.442 rows=33152 loops=95)
                                   Index Cond: ((project = me.project) AND (jobset = me.name))
                             ->  Bitmap Index Scan on indexbuildsonfinishedstoptime  (cost=0.00..377.38 rows=20614 width=0) (actual time=5.831..5.831 rows=119371 loops=94)
                                   Index Cond: (finished = 0)
         SubPlan 2
           ->  Aggregate  (cost=824.30..824.31 rows=1 width=8) (actual time=16.111..16.111 rows=1 loops=95)
                 ->  Index Scan using indexbuildsonjobsetiscurrent on builds a_1  (cost=0.56..823.66 rows=253 width=0) (actual time=0.052..15.847 rows=7005 loops=95)
                       Index Cond: ((project = me.project) AND (jobset = me.name))
                       Filter: ((buildstatus <> 0) AND (finished = 1))
                       Rows Removed by Filter: 26147
         SubPlan 3
           ->  Aggregate  (cost=825.29..825.30 rows=1 width=8) (actual time=16.627..16.627 rows=1 loops=95)
                 ->  Index Scan using indexbuildsonjobsetiscurrent on builds a_2  (cost=0.56..823.66 rows=652 width=0) (actual time=0.485..15.811 rows=25183 loops=95)
                       Index Cond: ((project = me.project) AND (jobset = me.name))
                       Filter: ((finished = 1) AND (buildstatus = 0))
                       Rows Removed by Filter: 7969
         SubPlan 4
           ->  Aggregate  (cost=333.01..333.02 rows=1 width=8) (actual time=10.545..10.545 rows=1 loops=95)
                 ->  Index Only Scan using indexbuildsonjobsetiscurrent on builds a_3  (cost=0.56..330.75 rows=905 width=0) (actual time=0.019..9.480 rows=33152 loops=95)
                       Index Cond: ((project = me.project) AND (jobset = me.name))
                       Heap Fetches: 1270929
 Planning Time: 0.655 ms
 JIT:
   Functions: 50
   Options: Inlining false, Optimization false, Expressions true, Deforming true
   Timing: Generation 5.782 ms, Inlining 0.000 ms, Optimization 0.922 ms, Emission 21.528 ms, Total 28.232 ms
 Execution Time: 5168.650 ms
(41 rows)
```

After:

```
 Sort  (cost=138217.71..138217.95 rows=95 width=534) (actual time=937.503..937.508 rows=95 loops=1)
   Sort Key: me.hidden, me.enabled DESC, me.name
   Sort Method: quicksort  Memory: 60kB
   ->  Seq Scan on jobsets me  (cost=0.00..138214.59 rows=95 width=534) (actual time=2.061..937.269 rows=95 loops=1)
         Filter: (project = 'nixpkgs'::text)
         Rows Removed by Filter: 313
         SubPlan 1
           ->  Aggregate  (cost=4.46..4.47 rows=1 width=8) (actual time=0.123..0.123 rows=1 loops=95)
                 ->  Index Only Scan using buildsjobsetidcurrentunfinished on builds a  (cost=0.42..4.45 rows=2 width=0) (actual time=0.002..0.084 rows=964 loops=95)
                       Index Cond: (jobset_id = me.id)
                       Heap Fetches: 0
         SubPlan 2
           ->  Aggregate  (cost=529.38..529.39 rows=1 width=8) (actual time=2.869..2.869 rows=1 loops=95)
                 ->  Index Only Scan using buildsjobsetidcurrentfinishedstatus on builds a_1  (cost=0.43..517.73 rows=4661 width=0) (actual time=1.901..2.589 rows=7005 loops=95)
                       Index Cond: (jobset_id = me.id)
                       Filter: (buildstatus <> 0)
                       Rows Removed by Filter: 25183
                       Heap Fetches: 0
         SubPlan 3
           ->  Aggregate  (cost=402.48..402.49 rows=1 width=8) (actual time=3.054..3.054 rows=1 loops=95)
                 ->  Index Only Scan using buildsjobsetidcurrentfinishedstatus on builds a_2  (cost=0.43..372.47 rows=12002 width=0) (actual time=0.007..2.115 rows=25183 loops=95)
                       Index Cond: ((jobset_id = me.id) AND (buildstatus = 0))
                       Heap Fetches: 0
         SubPlan 4
           ->  Aggregate  (cost=517.79..517.80 rows=1 width=8) (actual time=3.815..3.815 rows=1 loops=95)
                 ->  Index Only Scan using buildsjobsetidcurrent2 on builds a_3  (cost=0.43..476.12 rows=16668 width=0) (actual time=0.007..2.573 rows=33152 loops=95)
                       Index Cond: (jobset_id = me.id)
                       Heap Fetches: 0
 Planning Time: 0.483 ms
 Execution Time: 937.565 ms
(30 rows)
```

The migration takes about 9 minutes total on a replica of hydra.

### CODE REVIEW

File: src/lib/Hydra/Helper/Nix.pm [hunk 1]
```diff
@@ -115,10 +115,10 @@ sub jobsetOverview_ {
...
         { order_by => ["hidden ASC", "enabled DESC", "name"]
         , "+select" =>
-          [ "(select count(*) from Builds as a where a.finished = 0 and me.project = a.project and me.name = a.jobset and a.isCurrent = 1)"
-          , "(select count(*) from Builds as a where a.finished = 1 and me.project = a.project and me.name = a.jobset and buildstatus <> 0 and a.isCurrent = 1)"
-          , "(select count(*) from Builds as a where a.finished = 1 and me.project = a.project and me.name = a.jobset and buildstatus = 0 and a.isCurrent = 1)"
-          , "(select count(*) from Builds as a where me.project = a.project and me.name = a.jobset and a.isCurrent = 1)"
+          [ "(select count(*) from Builds as a where me.id = a.jobset_id and a.finished = 0 and a.isCurrent = 1)"
+          , "(select count(*) from Builds as a where me.id = a.jobset_id and a.finished = 1 and buildstatus <> 0 and a.isCurrent = 1)"
+          , "(select count(*) from Builds as a where me.id = a.jobset_id and a.finished = 1 and buildstatus = 0 and a.isCurrent = 1)"
+          , "(select count(*) from Builds as a where me.id = a.jobset_id and a.isCurrent = 1)"
```
Comment (grahamc) [inline src/lib/Hydra/Helper/Nix.pm:121]: :heavy_check_mark:

File: src/lib/Hydra/Controller/Build.pm [hunk 2]
```diff
@@ -83,8 +83,7 @@ sub build_GET {
 
     if ($build->finished) {
         $c->stash->{prevBuilds} = [$c->model('DB::Builds')->search(
-            { project => $c->stash->{project}->name
-            , jobset => $c->stash->{jobset}->name
+            { jobset_id => $c->stash->{jobset}->id
```
Comment (grahamc) [inline src/lib/Hydra/Controller/Build.pm]: Turns out this entire query is unused.

File: src/lib/Hydra/Helper/CatalystUtils.pm [hunk 3]
```diff
@@ -64,8 +64,7 @@ sub getNextBuild {
     (my $nextBuild) = $c->model('DB::Builds')->search(
       { finished => 1
       , system => $build->system
-      , project => $build->get_column('project')
-      , jobset => $build->get_column('jobset')
+      , jobset_id => $build->get_column('jobset_id')
```
Comment (grahamc) [inline src/lib/Hydra/Helper/CatalystUtils.pm:67]: :heavy_check_mark:

File: src/lib/Hydra/Helper/CatalystUtils.pm [hunk 4]
```diff
@@ -81,8 +80,7 @@ sub getPreviousSuccessfulBuild {
     (my $prevBuild) = $c->model('DB::Builds')->search(
       { finished => 1
       , system => $build->system
-      , project => $build->get_column('project')
-      , jobset => $build->get_column('jobset')
+      , jobset_id => $build->get_column('jobset_id')
```
Comment (grahamc) [inline src/lib/Hydra/Helper/CatalystUtils.pm:83]: :+1:

File: src/lib/Hydra/Controller/Channel.pm [hunk 5]
```diff
@@ -22,8 +22,7 @@ sub channel : Chained('/') PathPart('channel/custom') CaptureArgs(3) {
 
     my $lastSuccessful = $c->model('DB::Builds')->find(
         { 'eval.hasnewbuilds' => 1
-        , project => $projectName
-        , jobset => $jobsetName
+        , jobset_id => $c->stash->{jobset}->id,
```
Comment (grahamc) [inline src/lib/Hydra/Controller/Channel.pm:25]: :heavy_check_mark:

File: src/lib/Hydra/Controller/Build.pm [hunk 6]
```diff
@@ -81,26 +81,6 @@ sub build_GET {
...
-            }
-          , { join => "actualBuildStep"
-            , "+select" => ["actualBuildStep.stoptime - actualBuildStep.starttime"]
-            , "+as" => ["actualBuildTime"]
-            , order_by => "me.id DESC"
-            , rows => 50
-            }
-          )
-        ];
-    }
```
Comment (grahamc) [inline src/lib/Hydra/Controller/Build.pm:102]: Turns out this entire query is unused.

**Review by Ma27 — APPROVED** [review]
The diff looks okay, do you need somebody to test-deploy it? :)

**Comment by grahamc** [comment]
@Ma27 if you would do me the honor, that would be great!

**Comment by tomberek** [comment]
@grahamc it builds and runs, but my current hydra doesn't have enough jobs to measure the perf difference.

**Comment by Ma27** [comment]
@grahamc what @tomberek said also applies to me :)

**Comment by grahamc** [comment]
Great! Well, it is useful to check that I haven't outright *broken* anything :D. Thanks for checking!

**Comment by grahamc** [comment]
https://hydra.nixos.org/project/nixpkgs is loading in like 2.5s now :), down from 5-25s.

**Pull request merged**